Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -65,24 +83,53 @@ public final class SourceKitServer: LanguageServer {
     super.init(client: client)
   }
 
+  /// Registers the given `workspace` in the SourceKit server.
+  private func addWorkspace(_ workspace: Workspace) {
+    // If the workspace doesn't have a rootUri, use "". This will match it against any file, but will also make sure that any workspaces witha rootUri take precedence over this one.
+    _workspaces.append((workspace, workspace.rootUri?.canonicalizedPseudoPath ?? ""))
+  }
+
+  /// **Public for testing purposes**
+  public func _addWorkspace(_ workspace: Workspace) {
+    addWorkspace(workspace)
+  }
+
+  /// Returns the most specific workspace for the given `uri`.
+  /// This is done as follows:
+  ///  - The rootUri of the workspaces and `uri` are canonicalized to remove symlinks, '..' etc.
+  ///  - We select the workspace whose canonicalized rootUri is the longest prefix match to the canonicalized path of `uri`
+  /// This method returns `nil` if `uri` is in no open workspace.
+  public func workspaceForDocument(uri: DocumentURI) -> Workspace? {
+    let canonicalizedFilePath = uri.canonicalizedPseudoPath
+    var bestWorkspaceMatch: (workspace: Workspace, canonicalizedRootPath: String)? = nil
+    for (workspace, canonicalizedWorkspaceRootPath) in _workspaces where canonicalizedFilePath.hasPrefix(canonicalizedWorkspaceRootPath) {
```

Comment:
If there's only one workspace, should we maintain the old behavior? I think we would regress in workspaces that give a rootUri but can the user could open files outside of there (e.g. derived files, etc)

Reply:
Also not sure how safe this is to generalize - for SwiftPM you might be able to have multiple BuildSystem instances, but for a monorepo with a unified build system you can really only have one

Reply:
> I was thinking the same but decided against it because I thought it was confusing if opening a second workspace changed the way that files were matched to these workspaces. I also assumed that opening files outside of that rootUri isn’t common, but maybe that assumption is wrong. I wouldn’t be terribly opposed to matching all files to a workspace if only one is open.

Yeah, it seems like we need a way to explicitly force the one-workspace behavior. Is it possible that `workspaceFolders` are added/removed after initialization, I think so, right? So in that case maybe a flag to force the old behavior is OK?

> Would issues would you expect here? Having multiple build system instances for unrelated projects seems like it should be possible to me. And for a monorepo I would expect that the rootUri is the base of the monorepo? Or am I making wrong assumptions here?

It's possible that the user has multiple folders of the monorepo open (2 `workspaceFolders`), e.g. `llvm-project/clang` and `llvm-project/clang-tools-extra`, but the root-uri we'd want to use is `llvm-project` and only share a single instance, not have two. IIUC correctly, with this implementation as is, we have no way to support this use case, right?

Reply:
> Yeah, it seems like we need a way to explicitly force the one-workspace behavior. Is it possible that `workspaceFolders` are added/removed after initialization, I think so, right? So in that case maybe a flag to force the old behavior is OK?

Sleeping over it, I think maintaining the current behavior if only a single workspace is open is reasonable. If you only have one workspace, there’s no reason to assume that a file doesn’t belong to that workspace.

I also thought about adding a similar fallback logic for multiple workspaces but I think falling back to the first or last workspace will be surprising.

> It's possible that the user has multiple folders of the monorepo open (2 `workspaceFolders`), e.g. `llvm-project/clang` and `llvm-project/clang-tools-extra`, but the root-uri we'd want to use is `llvm-project` and only share a single instance, not have two. IIUC correctly, with this implementation as is, we have no way to support this use case, right?

Hmm, this really makes things tricky because clangd doesn’t support multiple workspace folders. I can think of multiple options
1. We refuse to open files from two different workspace folders in clangd. That is, you open a file from workspace folder A, it succeeds, you open a file from workspace folder B and it fails with an error because workspace folder A is already open in clangd and clangd can’t handle multiple workspace folders. This doesn’t support your use case but your use case isn’t supported today either, so that shouldn’t be a regression.
2. Create a separate clangd instance for the different workspace folders. I am not entirely sure how well multiple clangd instance can co-exist on the same project, do you have any experience with that @DavidGoldman? If there’s no issue, I think this would be a very elegant solution.
3. Have a single `clangd` instance but don’t open it with the rootUri of any particular workspace, but with the deprecated `rootUri` property of the `initialize` request. This would support your use case but I really hate this option because it breaks the assumption that the deprecated `rootUri` property in the initialize request is going to be ignored if you pass `workspaceFolders`

Reply:
> I think we would regress in workspaces that give a rootUri but the user could open files outside of there (e.g. derived files, etc

This does seem important to me. Files in a build directory are either going to be opened directly if the editor doesn't know anything about them, or else the editor might want to add a root directory for the build directory, and in either case the correct interpretation is to consider them part of the same overall workspace.  If you have multiple build systems across workspace roots, that means we need to figure out which one the file belongs to to the best of our ability.  This seems kind of orthogonal from workspace roots anyway, since you could have multiple build systems in a single root.  I wonder if we should keep a single `Workspace` and have the build system layer deal with the multi-root support.